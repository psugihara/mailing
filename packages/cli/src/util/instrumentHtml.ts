import { parse } from "node-html-parser";

export default function instrumentHtml({
  html,
  sendId,
  apiUrl,
}: {
  html: string;
  sendId: string;
  apiUrl: string;
}) {
  // find the a, get href, calculate new url, replace base 64 encoded href
  const root = parse(html);
  const links = root.querySelectorAll("a");
  for (const link of links) {
    const href = link.getAttribute("href");
    if (!href) continue;
    const url = new URL("/api/click", apiUrl);
    url.searchParams.set("sendId", sendId);
    url.searchParams.set("url", href);
    link.setAttribute("href", url.toString());
  }

  // add <img alt="" src="apiUrl?t=BASE_64" /> to end of body
  const body = root.querySelector("body");
  if (body) {
    const img = parse(
      `<img alt="" src="${apiUrl}/api/open?sendId=${sendId}" />`
    );
    body.appendChild(img);
  } else {
    console.warn("Could not find body in html");
  }

  return root.toString();
}
