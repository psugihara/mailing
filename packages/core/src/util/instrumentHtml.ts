import { parse } from "node-html-parser";

export default function instrumentHtml({
  html,
  messageId,
  apiUrl,
}: {
  html: string;
  messageId: string;
  apiUrl: string;
}) {
  // find the a, get href, calculate new url, replace href
  const root = parse(html);
  const links = root.querySelectorAll("a");
  for (const link of links) {
    const href = link.getAttribute("href");
    if (!href) continue;
    const url = new URL("/api/hooks/click", apiUrl);
    url.searchParams.set("messageId", messageId);
    url.searchParams.set("url", href);
    link.setAttribute("href", url.toString());
  }

  // add open tracking pixel to end of body
  const body = root.querySelector("body");
  if (body) {
    const img = parse(
      `<img alt="" src="${apiUrl}/api/hooks/open?messageId=${messageId}" />`
    );
    body.appendChild(img);
  } else {
    throw new Error("no body found in html");
  }

  return root.toString();
}
