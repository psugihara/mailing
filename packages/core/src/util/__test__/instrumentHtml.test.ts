import instrumentHtml from "../instrumentHtml";

describe("instrumentHtml", () => {
  const messageId = "123";
  const apiUrl = "https://api.com";

  it("should instrument html", () => {
    const html = `
    <html>
      <head>
        <title>Test</title>
      </head>
      <body>
        <a href="https://google.com">Google</a>
        <a href="https://yahoo.com">Yahoo</a>
      </body>
    </html>
  `;

    const result = instrumentHtml({ html, messageId, apiUrl });

    expect(result).toContain(
      'href="https://api.com/api/hooks/click?messageId=123&url=https%3A%2F%2Fgoogle.com"'
    );
    expect(result).not.toContain('href="https://google.com"');
    expect(result).toContain(
      '<img alt="" src="https://api.com/api/hooks/open?messageId=123" >'
    );

    expect(result).toMatchSnapshot();
  });

  it("throws error if no body found", () => {
    const html = `
      <html>
        <head>
          <title>Test</title>
        </head>
      </html>
    `;

    expect(() => instrumentHtml({ html, messageId, apiUrl })).toThrow();
  });
});
