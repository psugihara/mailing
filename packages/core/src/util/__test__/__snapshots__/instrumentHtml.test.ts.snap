// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instrumentHtml should instrument html 1`] = `
"
    <html>
      <head>
        <title>Test</title>
      </head>
      <body>
        <a href="https://api.com/api/hooks/click?messageId=123&url=https%3A%2F%2Fgoogle.com">Google</a>
        <a href="https://api.com/api/hooks/click?messageId=123&url=https%3A%2F%2Fyahoo.com">Yahoo</a>
      <img alt="" src="https://api.com/api/hooks/open?messageId=123" ></body>
    </html>
  "
`;
